File: app/src/main/java/messaging/capaxit/nl/fcmmessaging/MainActivity.java
package messaging.capaxit.nl.fcmmessaging;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.View;
import android.widget.EditText;
import com.google.firebase.iid.FirebaseInstanceId;
import com.google.firebase.messaging.FirebaseMessaging;
import rx.Completable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

import java.io.IOException;

public class MainActivity extends AppCompatActivity {
    private static final String TAG = MainActivity.class.getSimpleName();
    private EditText topicName;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        topicName = (EditText) findViewById(R.id.topicName);
        findViewById(R.id.subsribeToTopic).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View v) {
                subscribe();
            }
        });
        findViewById(R.id.unsubscribeFromTopic).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View v) {
                unsubscribe();
            }
        });
        findViewById(R.id.registerToFcm).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View v) {
                registerToFcm();
            }
        });
        findViewById(R.id.unregisterFromFcm).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(final View v) {
                unregisterFromFcm();
            }
        });
        findViewById(R.id.logFcmToken).setOnClickListener(new View.OnClickListener() {
            public void onClick(View view) {
                Log.i(TAG, FirebaseInstanceId.getInstance().getToken());
            }
        });
    }

    private void unregisterFromFcm() {
        getUnregisterObservable()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(throwable -> Log.e(TAG, "Error", throwable), () -> Log.i(TAG, "Unregistered"));
    }

    @NonNull
    private Completable getUnregisterObservable() {
        return Completable.create(s -> {
            try {
                FirebaseInstanceId.getInstance().deleteInstanceId();
                s.onCompleted();
            } catch (IOException e) {
                s.onError(e);
            }
        });
    }

    private void registerToFcm() {
        FirebaseInstanceId.getInstance().getToken();
    }

    private void unsubscribe() {
        FirebaseMessaging.getInstance().unsubscribeFromTopic(topicName.getText().toString());
    }

    private void subscribe() {
        FirebaseMessaging.getInstance().subscribeToTopic(topicName.getText().toString());
    }
}

File: app/src/main/java/messaging/capaxit/nl/fcmmessaging/MainActivity.kt
package messaging.capaxit.nl.fcmmessaging

import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.util.Log
import com.google.firebase.iid.FirebaseInstanceId
import com.google.firebase.messaging.FirebaseMessaging
import kotlinx.android.synthetic.main.activity_main.*
import rx.Completable
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers
import java.io.IOException

class MainActivity : AppCompatActivity() {
    private val unregisterObservable: Completable
        get() = Completable.create { s ->
            try {
                FirebaseInstanceId.getInstance().deleteInstanceId()
                s.onCompleted()
            } catch (e: IOException) {
                s.onError(e)
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        MessagingChannels.createMessagingChannels(this)
        setContentView(R.layout.activity_main)
        subsribeToTopic.setOnClickListener { subscribe() }
        unsubscribeFromTopic.setOnClickListener { unsubscribe() }
        registerToFcm.setOnClickListener { registerToFcm() }
        unregisterFromFcm.setOnClickListener { unregisterFromFcm() }
        logFcmToken.setOnClickListener { Log.i(TAG, FirebaseInstanceId.getInstance().token) }
    }

    private fun unregisterFromFcm() {
        unregisterObservable
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ throwable -> Log.e(TAG, "Error", throwable) }, { Log.i(TAG, "Unregistered") })
    }

    private fun registerToFcm() {
        FirebaseInstanceId.getInstance().instanceId.addOnSuccessListener {
            Log.w(TAG, it.token)
        }
    }

    private fun unsubscribe() {
        FirebaseMessaging.getInstance().unsubscribeFromTopic(topicName.text.toString())
    }

    private fun subscribe() {
        FirebaseMessaging.getInstance().subscribeToTopic(topicName.text.toString())
    }

    companion object {
        private val TAG = MainActivity::class.java.simpleName
    }
}
